File: scripts/platform/slice-collisions.py
#!/usr/bin/env python3
"""Compute the maximum FILE-DISJOINT set of platform tickets that can run concurrently.

The parallelism limit on this program is not disk and not CPU — it is merge conflicts.
Worktrees make concurrent edits *safe* (no clobbering) but do nothing to prevent two agents
editing the same file and colliding at merge. That is a mechanical property of the `owns`
column in docs/platform/wave_plan.tsv, so it is computed here rather than eyeballed.

  python3 scripts/platform/slice-collisions.py                 # max concurrent set from the next wave
  python3 scripts/platform/slice-collisions.py T-190 T-191     # what may JOIN those already in flight
  python3 scripts/platform/slice-collisions.py --repack        # rebuild wave_plan.tsv from the registry
  python3 scripts/platform/slice-collisions.py --check T-190   # is T-190 safe against everything running?

Unlike the T-181 mod version this takes the plan path from TBD_WAVE_PLAN, so the same logic
serves any program. Default is the platform plan.
"""
import csv, json, os, sys, subprocess, collections

ROOT = subprocess.run(['git', 'rev-parse', '--show-toplevel'],
                      capture_output=True, text=True).stdout.strip()
PLAN = os.environ.get('TBD_WAVE_PLAN', os.path.join(ROOT, 'docs/platform/wave_plan.tsv'))
REGISTRY = os.path.join(ROOT, '.ai/tickets/registry.json')
# Integration attention, not disk, is the real ceiling: every agent returns a dense report the
# command center must actually read. Measured on T-181: three was far too low, twenty is too many
# to integrate in one sitting. Eight is the working compromise — raise it if you are keeping up.
MAX_CONCURRENT = int(os.environ.get('TBD_MAX_CONCURRENT', '8'))

# Ordering constraints that file-disjointness cannot express. Each of these is a case where two
# tickets touch DIFFERENT files but one still has to land first, so the collision computation alone
# would happily run them together and produce a broken tree.
#
#   T-273 -> T-237 -> T-238  `ticket check` is inside the wave gate. T-237 wires it to validate
#                            against schema.json, and schema.json is a month stale — every one of
#                            the 113 tickets violates it today. Land T-237 first and the gate goes
#                            red on the whole registry, failing every subsequent wave.
#   T-241 -> zones consumers The doc has no `zones` root at all. Four tickets would each invent a
#                            different one; T-241 declares the vocabulary once.
#   T-222 -> sync consumers  CLIENT_ID is hardcoded to 1. Any sync transport that lands first
#                            corrupts documents on every multi-peer merge.
#   T-257 -> undo consumers  `objectives`/`markers` are cleared by hydrate but not undo-scoped, so
#                            both features would ship non-undoable.
#   T-186 -> T-209 -> T-251  test lane, then CI wiring, then deploy.
#   T-290 LAST               it annotates fields as non-consumed that five earlier tickets build.
DEPS = {
    'T-237': ['T-273'], 'T-238': ['T-273', 'T-237'],
    'T-201': ['T-241'], 'T-211': ['T-241'], 'T-212': ['T-241', 'T-257'], 'T-275': ['T-241'],
    'T-190': ['T-222'], 'T-295': ['T-222'], 'T-213': ['T-257'],
    'T-209': ['T-186'], 'T-251': ['T-209'],
}
RUN_LAST = {'T-290'}


def plan_rows():
    if not os.path.exists(PLAN):
        sys.exit(f"no wave plan at {PLAN} (set TBD_WAVE_PLAN)")
    with open(PLAN) as fh:
        for r in csv.reader(fh, delimiter='\t'):
            if not r or r[0].startswith('#') or r[0] == 'wave':
                continue
            if len(r) < 4:
                continue
            yield {'wave': r[0], 'id': r[1], 'title': r[2],
                   'owns': [p.strip() for p in r[3].split(';') if p.strip()]}


def registry():
    with open(REGISTRY) as fh:
        return {t['id']: t for t in json.load(fh)['tickets']}


def shipped(tid, reg):
    return reg.get(tid, {}).get('status') in ('shipped', 'cancelled')


def dispatchable(tid, reg):
    """Can a slice AGENT take this ticket, or is a human the only one who can?

    Two ways a ticket is undispatchable even though it is not shipped:

      status `deferred`  — a slice agent already took it and refused with cause. T-205 and T-206
                           are the live case: the missing vehicle/item data only exists behind a
                           Workbench export pass. Re-dispatching burns a whole agent to re-derive
                           the same refusal.
      executor != claude-code — the D5 executor gate in CLAUDE.md. `workbench`, `human` and `ci`
                           rows are operator work by definition.

    Without this, `pack()` filtered on shipped/cancelled ALONE and kept offering both tickets at
    the head of every dispatch set, where they would have consumed 2 of 8 slots per wave forever.
    """
    t = reg.get(tid, {})
    if t.get('status') in ('shipped', 'cancelled', 'deferred', 'blocked'):
        return False
    return t.get('executor', 'claude-code') == 'claude-code'


def collides(a, b):
    """Two tickets collide if any owned path overlaps — including prefix containment,
    so `apps/website/api/src/` collides with `apps/website/api/src/handlers/admin.rs`."""
    for x in a:
        for y in b:
            if x == y or x.startswith(y.rstrip('/') + '/') or y.startswith(x.rstrip('/') + '/'):
                return True
    return False


def pack(cands, already=(), landed=None, enforce_deps=True):
    """Greedy maximum disjoint set, honouring plan order (which is priority order) and DEPS.

    `landed` defaults to everything already shipped. It MUST NOT default to an empty set: repack()
    seeded it explicitly but main() did not, so `wave.sh prep` — the only dispatch view — silently
    skipped every ticket carrying a DEPS edge, forever. 11 tickets were unreachable, including
    T-209 whose dependency T-186 had already shipped. Computing it here covers both callers.
    """
    if landed is None:
        landed = {tid for tid, t in registry().items()
                  if t.get('status') in ('shipped', 'cancelled')}
    chosen, used = [], list(already)
    for c in cands:
        if enforce_deps:
            if c['id'] in RUN_LAST:
                continue
            if any(d not in landed for d in DEPS.get(c['id'], ())):
                continue
        if any(collides(c['owns'], u) for u in used):
            continue
        chosen.append(c)
        used.append(c['owns'])
        if len(chosen) + len(already) >= MAX_CONCURRENT:
            break
    return chosen


def repack():
    """Rebuild the plan from the registry, re-packing every unshipped ticket by disjointness.
    Preserves each row's `owns` — only the wave numbers move."""
    reg = registry()
    rows = [r for r in plan_rows() if dispatchable(r['id'], reg)]
    done = [r for r in plan_rows() if not dispatchable(r['id'], reg)]
    waves, remaining, n = [], list(rows), 0
    # Seed `landed` with everything already shipped. Without this, a DEPS edge pointing at a shipped
    # ticket can never be satisfied — the dependency is filtered out of `rows` as done, so it never
    # enters `landed`, and every dependent deadlocks. Hit for real on 2026-07-26 once T-186 shipped:
    # T-209 -> T-186 and T-251 -> T-209 both became unschedulable.
    landed = {tid for tid, t in registry().items()
              if t.get('status') in ('shipped', 'cancelled')}
    last = [r for r in remaining if r['id'] in RUN_LAST]
    remaining = [r for r in remaining if r['id'] not in RUN_LAST]
    while remaining:
        n += 1
        w = pack(remaining, landed=landed)
        if not w:
            # Everything left is either colliding or dep-blocked. Take the first whose deps are
            # satisfied; if none are, the DEPS table has a cycle and that is a bug worth shouting about.
            free = [r for r in remaining if all(d in landed for d in DEPS.get(r['id'], ()))]
            if not free:
                sys.exit(f"DEPS deadlock: {[r['id'] for r in remaining][:8]} — check the DEPS table")
            w = [free[0]]
        for r in w:
            remaining.remove(r)
            landed.add(r['id'])
        waves.append(w)
    for r in last:                     # RUN_LAST tickets get their own trailing wave
        waves.append([r])
    out = ["# Platform wave plan — WHICH tickets run together, and in what order.",
           "# Columns: wave <TAB> ticket <TAB> title <TAB> owns (semicolon-separated paths)",
           "# Waves are packed by FILE-DISJOINTNESS in priority order.",
           "# Regenerate: python3 scripts/platform/slice-collisions.py --repack", "#"]
    for r in done:
        out.append(f"0\t{r['id']}\t{r['title']}\t{'; '.join(r['owns'])}")
    for i, w in enumerate(waves, 1):
        for r in w:
            out.append(f"{i}\t{r['id']}\t{r['title']}\t{'; '.join(r['owns'])}")
    with open(PLAN, 'w') as fh:
        fh.write('\n'.join(out) + '\n')
    print(f"repacked {sum(len(w) for w in waves)} open tickets into {len(waves)} waves "
          f"({len(done)} already shipped, parked at wave 0)")


def main():
    args = [a for a in sys.argv[1:] if not a.startswith('--')]
    flags = {a for a in sys.argv[1:] if a.startswith('--')}
    if '--repack' in flags:
        return repack()

    reg = registry()
    rows = [r for r in plan_rows() if dispatchable(r['id'], reg)]
    by_id = {r['id']: r for r in rows}

    if '--check' in flags:
        if not args:
            sys.exit("--check needs a ticket id")
        t = by_id.get(args[0])
        if not t:
            sys.exit(f"{args[0]} is not an open ticket in {os.path.relpath(PLAN, ROOT)}")
        bad = [o['id'] for o in rows if o['id'] != t['id'] and collides(t['owns'], o['owns'])]
        print(f"{t['id']} owns: {'; '.join(t['owns'])}")
        print("collides with:", ', '.join(bad) if bad else "nothing — safe to run alongside anything")
        return

    running = [by_id[a] for a in args if a in by_id]
    for a in args:
        if a not in by_id:
            print(f"warning: {a} is not an open ticket in the plan", file=sys.stderr)

    cands = [r for r in rows if r['id'] not in {x['id'] for x in running}]
    picked = pack(cands, already=[r['owns'] for r in running])

    if running:
        print(f"already in flight ({len(running)}):")
        for r in running:
            print(f"  {r['id']:8s} {r['title'][:60]}")
        print(f"\nmay join them ({len(picked)}, cap {MAX_CONCURRENT}):")
    else:
        nxt = min((r['wave'] for r in rows), key=lambda w: int(w))
        print(f"next wave is {nxt}. Max disjoint dispatch set ({len(picked)}, cap {MAX_CONCURRENT}):")
    for r in picked:
        print(f"  {r['id']:8s} {r['title'][:60]}")
        print(f"           owns: {'; '.join(r['owns'])}")
    if not picked:
        print("  (none — everything left collides with what is already running)")

    blocked = collections.Counter()
    for c in cands:
        if c not in picked:
            for r in picked + running:
                if collides(c['owns'], r['owns']):
                    blocked[r['id']] += 1
    if blocked:
        print("\nmost-contended tickets (blocking the most others):")
        for tid, n in blocked.most_common(5):
            print(f"  {tid} blocks {n}")


if __name__ == '__main__':
    main()

File: xtask/src/slice_collisions.rs
//! T-620 — the maximum FILE-DISJOINT set of platform tickets that can run concurrently.
//!
//! Originally a byte-for-byte port of `scripts/platform/slice-collisions.py` over the wave-plan
//! TSV; since T-912.2 the plan is the compiled `.ai/tickets/wave.lock` and the packing facts —
//! `owns`, `depends_on`, `pack_last` (T-912.1) — live on the tickets, so this command reads the
//! lock plus the ticket files and nothing else. The TSV, its plan-path env override, and the
//! label-format checks that guarded a hand-kept column (T-616/T-623 F5) died with it: lock wave
//! numbers are typed integers written by one writer, and a missing lock is a DidNotRun refusal
//! from [`crate::wave_lock::load`], never an empty dispatch set.
//!
//! The parallelism limit on this program is not disk and not CPU — it is merge conflicts.
//! Worktrees make concurrent edits *safe* (no clobbering) but do nothing to prevent two agents
//! editing the same file and colliding at merge. That is a mechanical property of each ticket's
//! `owns` field, so it is computed here rather than eyeballed.
//!
//!   cargo xtask slice-collisions                 # max concurrent set from the open waves
//!   cargo xtask slice-collisions T-190 T-191     # what may JOIN those already in flight
//!   cargo xtask slice-collisions --repack        # alias for `cargo xtask wave repack`
//!   cargo xtask slice-collisions --check T-190   # is T-190 safe against everything running?
//!
//! `--repack` is an ALIAS, not a second writer: the lock has exactly one compiler
//! ([`crate::wave_lock::cmd_repack`]), and this spelling survives only because a generation of
//! runbooks and muscle memory says `slice-collisions --repack` when the plan is stale.

use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use tbd_tickets::{Ticket, parse_ticket_toml};

use crate::wave_lock;

/// Integration attention, not disk, is the real ceiling: every agent returns a dense report the
/// command center must actually read. Measured on T-181: three was far too low, twenty is too many
/// to integrate in one sitting. Eight is the working compromise — raise it if you are keeping up.
fn max_concurrent() -> usize {
    wave_lock::max_concurrent()
}

/// Per-ticket packing facts — `owns`, `depends_on`, `pack_last` — read from EVERY
/// `.ai/tickets/T-*.toml`, children included.
///
/// Until T-912.1 the ordering constraints that file-disjointness cannot express (two tickets touch
/// DIFFERENT files but one must land first) were a hardcoded 11-row dependency table plus a
/// run-last list in this file, and `owns` lived only in the wave-plan TSV column. All three are
/// ticket fields now, and T-912.2 compiled the wave labels into the lock.
///
/// NOT the parents-only registry loader: `load_phase2_tree` walks PARENT files only, so child ids
/// like T-181.23 or T-090.4 would be absent from its map. Glob the directory directly.
struct TicketFacts {
    owns: HashMap<String, Vec<String>>,
    depends_on: HashMap<String, Vec<String>>,
    pack_last: HashSet<String>,
}

impl TicketFacts {
    fn deps_of(&self, id: &str) -> &[String] {
        self.depends_on.get(id).map(Vec::as_slice).unwrap_or(&[])
    }
}

fn ticket_facts(root: &Path) -> Result<TicketFacts> {
    let dir = crate::tickets_store::tickets_dir(root);
    let mut facts = TicketFacts {
        owns: HashMap::new(),
        depends_on: HashMap::new(),
        pack_last: HashSet::new(),
    };
    for ent in std::fs::read_dir(&dir).with_context(|| dir.display().to_string())? {
        let ent = ent?;
        let name = ent.file_name();
        let name = name.to_string_lossy();
        if !name.starts_with("T-") || !name.ends_with(".toml") {
            continue;
        }
        let text = std::fs::read_to_string(ent.path())?;
        let t = parse_ticket_toml(&text)
            .map_err(|e| anyhow::anyhow!("{}: {e}", ent.path().display()))?;
        let (id, owns, depends_on, pack_last) = match t {
            Ticket::Program(p) => (p.id, p.owns, p.depends_on, p.pack_last),
            Ticket::Work(w) => (w.id, w.owns, w.depends_on, w.pack_last),
        };
        if !owns.is_empty() {
            facts.owns.insert(id.clone(), owns);
        }
        if !depends_on.is_empty() {
            facts.depends_on.insert(id.clone(), depends_on);
        }
        if pack_last == Some(true) {
            facts.pack_last.insert(id);
        }
    }
    Ok(facts)
}

#[derive(Clone, Debug)]
struct Row {
    wave: u32,
    id: String,
    title: String,
    owns: Vec<String>,
}

fn repo_root() -> Result<PathBuf> {
    let out = std::process::Command::new("git")
        .args(["rev-parse", "--show-toplevel"])
        .output()
        .context("git rev-parse --show-toplevel")?;
    Ok(PathBuf::from(
        String::from_utf8_lossy(&out.stdout).trim().to_string(),
    ))
}

/// One row per lock entry, in lock order — wave 0 included, because `--check` and the collision
/// counters reason over the whole plan while the dispatch set filters to the open waves.
fn lock_rows(
    lock: &wave_lock::WaveLock,
    views: &HashMap<String, wave_lock::TicketView>,
    facts: &TicketFacts,
) -> Vec<Row> {
    let mut out = Vec::new();
    for w in &lock.waves {
        for id in &w.tickets {
            out.push(Row {
                wave: w.n,
                id: id.clone(),
                title: views.get(id).map(|v| v.title.clone()).unwrap_or_default(),
                owns: facts.owns.get(id).cloned().unwrap_or_default(),
            });
        }
    }
    out
}

/// Two tickets collide if any owned path overlaps — including prefix containment, so
/// `apps/website/api/src/` collides with `apps/website/api/src/handlers/admin.rs`.
fn collides(a: &[String], b: &[String]) -> bool {
    wave_lock::collides(a, b)
}

/// Greedy maximum disjoint set, honouring lock order (which is priority order) and ticket
/// `depends_on` / `pack_last`.
///
/// `blocking` is the set of dependency targets that can still land — open dispatchable tickets
/// not yet landed. A dep outside it is history (shipped/cancelled) or unschedulable
/// (idea/deferred/executor-gated), and the compiler already warned about the latter at repack;
/// treating those as blockers here would hide the very tickets the lock schedules.
fn pack<'r>(
    cands: &[&'r Row],
    already: &[Vec<String>],
    blocking: &HashSet<String>,
    facts: &TicketFacts,
    max: usize,
) -> Vec<&'r Row> {
    let mut chosen: Vec<&Row> = Vec::new();
    let mut used: Vec<Vec<String>> = already.to_vec();
    for c in cands {
        if facts.pack_last.contains(&c.id) {
            continue;
        }
        if facts.deps_of(&c.id).iter().any(|d| blocking.contains(d)) {
            continue;
        }
        if used.iter().any(|u| collides(&c.owns, u)) {
            continue;
        }
        chosen.push(c);
        used.push(c.owns.clone());
        if chosen.len() + already.len() >= max {
            break;
        }
    }
    chosen
}

/// Python `str[:n]` sliced by CHARACTER — these titles are full of em-dashes (3 bytes each), so
/// `&s[..60]` would both truncate differently and panic on a non-boundary index.
fn chars(s: &str, n: usize) -> String {
    s.chars().take(n).collect()
}

/* ─────────────────────────── unplanned-ticket warning ─────────────────────────── */

/// Dispatchable work tickets with NO row in the lock's waves 1+ — invisible to every dispatch
/// set this command computes.
///
/// THIS IS THE HOLE THAT MATTERS MOST HERE. Measured 2026-07-26 on the TSV ancestor of this
/// check: 15 of 42 open platform tickets — 36% of the backlog, including a P0 that broke all
/// production telemetry — were absent from every "Max disjoint dispatch set" this tool
/// confidently printed. The TSV-era check keyed on `program == "platform"` registry rows; the
/// lock-era rule is simpler and total: EVERY dispatchable work ticket must appear in waves 1+.
///
/// Deliberately a LOUD WARNING and not a hard exit — but unlike the TSV era, the fix is now one
/// command, because `wave repack` packs every open ticket from its own `owns`.
fn warn_unplanned(views: &HashMap<String, wave_lock::TicketView>, lock: &wave_lock::WaveLock) {
    let planned: HashSet<String> = lock.open_ids().into_iter().collect();
    let mut miss: Vec<&wave_lock::TicketView> = views
        .values()
        .filter(|v| v.dispatchable() && !planned.contains(&v.id))
        .collect();
    if miss.is_empty() {
        return;
    }
    miss.sort_by(|a, b| a.id.cmp(&b.id));
    eprintln!(
        "\n\x1b[33m! {} DISPATCHABLE TICKET(S) ARE NOT IN THE LOCK'S OPEN WAVES and cannot be dispatched:\x1b[0m",
        miss.len()
    );
    for v in &miss {
        eprintln!("    {:<10} {}", v.id, chars(&v.title, 58));
    }
    eprintln!("  Run `cargo xtask wave repack` — the compiler packs every open ticket.");
}

/* ─────────────────────────── entry point ─────────────────────────── */

pub fn run(argv: &[String]) -> Result<u8> {
    let root = repo_root()?;
    let max = max_concurrent();

    let args: Vec<&String> = argv.iter().filter(|a| !a.starts_with("--")).collect();
    let flags: HashSet<&str> = argv
        .iter()
        .filter(|a| a.starts_with("--"))
        .map(String::as_str)
        .collect();

    // One writer. This alias exists for runbook muscle memory only (see the module header).
    if flags.contains("--repack") {
        return wave_lock::cmd_repack(&root);
    }

    let lock = wave_lock::load(&root)?; // missing lock = DidNotRun refusal, never an empty set
    let views: HashMap<String, wave_lock::TicketView> = wave_lock::load_views(&root)?
        .into_iter()
        .map(|v| (v.id.clone(), v))
        .collect();
    let facts = ticket_facts(&root)?;
    let all = lock_rows(&lock, &views, &facts);

    // Open rows = the lock's waves 1+. Drift between the lock and the tickets is `wave check`'s
    // job (wired into `ticket check`); this command trusts the committed plan it was handed.
    let rows: Vec<&Row> = all.iter().filter(|r| r.wave > 0).collect();
    let by_id: HashMap<&str, &Row> = rows.iter().map(|r| (r.id.as_str(), *r)).collect();

    if flags.contains("--check") {
        let Some(want) = args.first() else {
            bail!("--check needs a ticket id");
        };
        let Some(t) = by_id.get(want.as_str()) else {
            bail!("{want} is not an open ticket in {}", wave_lock::LOCK_REL);
        };
        let bad: Vec<&str> = rows
            .iter()
            .filter(|o| o.id != t.id && collides(&t.owns, &o.owns))
            .map(|o| o.id.as_str())
            .collect();
        println!("{} owns: {}", t.id, t.owns.join("; "));
        println!(
            "collides with: {}",
            if bad.is_empty() {
                "nothing — safe to run alongside anything".to_string()
            } else {
                bad.join(", ")
            }
        );
        return Ok(0);
    }

    let mut running: Vec<&Row> = Vec::new();
    for a in &args {
        match by_id.get(a.as_str()) {
            Some(r) => running.push(r),
            None => eprintln!("warning: {a} is not an open ticket in the lock"),
        }
    }
    let running_ids: HashSet<&str> = running.iter().map(|r| r.id.as_str()).collect();
    let cands: Vec<&Row> = rows
        .iter()
        .copied()
        .filter(|r| !running_ids.contains(r.id.as_str()))
        .collect();
    let already: Vec<Vec<String>> = running.iter().map(|r| r.owns.clone()).collect();
    // Deps that can still block a candidate: open tickets not named as already running.
    let blocking: HashSet<String> = cands.iter().map(|r| r.id.clone()).collect();
    let picked = pack(&cands, &already, &blocking, &facts, max);

    if !running.is_empty() {
        println!("already in flight ({}):", running.len());
        for r in &running {
            println!("  {:<8} {}", r.id, chars(&r.title, 60));
        }
        println!("\nmay join them ({}, cap {max}):", picked.len());
    } else if rows.is_empty() {
        // A lock with an empty open set is the factory FINISHING, not the factory breaking —
        // the missing-lock refusal above is the failure case. Exit 0, but SAY SO.
        println!(
            "no open tickets in {} — every planned ticket is parked at wave 0. Nothing to dispatch.",
            wave_lock::LOCK_REL
        );
        warn_unplanned(&views, &lock);
        return Ok(0);
    } else {
        let nxt = rows.iter().map(|r| r.wave).min().unwrap_or(1);
        println!(
            "next wave is {nxt}. Max disjoint dispatch set ({}, cap {max}):",
            picked.len()
        );
    }
    for r in &picked {
        println!("  {:<8} {}", r.id, chars(&r.title, 60));
        println!("           owns: {}", r.owns.join("; "));
    }
    if picked.is_empty() {
        println!("  (none — everything left collides with what is already running)");
    }

    // Counter + most_common(5): count descending, ties by FIRST-INSERTION order.
    let picked_ids: HashSet<&str> = picked.iter().map(|r| r.id.as_str()).collect();
    let mut seen: Vec<&str> = Vec::new();
    let mut counts: HashMap<&str, u64> = HashMap::new();
    for c in &cands {
        if picked_ids.contains(c.id.as_str()) {
            continue;
        }
        for r in picked.iter().chain(running.iter()) {
            if collides(&c.owns, &r.owns) {
                let e = counts.entry(r.id.as_str()).or_insert_with(|| {
                    seen.push(r.id.as_str());
                    0
                });
                *e += 1;
            }
        }
    }
    if !counts.is_empty() {
        let mut ranked: Vec<(&str, u64)> = seen.iter().map(|id| (*id, counts[id])).collect();
        ranked.sort_by(|a, b| b.1.cmp(&a.1)); // stable => ties keep insertion order
        println!("\nmost-contended tickets (blocking the most others):");
        for (id, n) in ranked.iter().take(5) {
            println!("  {id} blocks {n}");
        }
    }

    warn_unplanned(&views, &lock);
    Ok(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn worktree_root() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .expect("xtask parent = repo/worktree root")
            .to_path_buf()
    }

    /// T-912.1: the hardcoded dependency/run-last tables must not come back — the packer reads
    /// ticket `depends_on` / `pack_last` / `owns`. The needle is assembled at runtime so this
    /// test's own source cannot satisfy the scan it performs.
    #[test]
    fn hardcoded_dep_tables_stay_deleted() {
        let src = include_str!("slice_collisions.rs");
        for name in ["DEPS", "RUN_LAST"] {
            let needle = format!("const {name}");
            assert!(
                !src.contains(&needle),
                "`{needle}` is back in slice_collisions.rs — T-912.1 moved it onto the tickets"
            );
        }
    }

    /// The packing facts come from the ticket files, children included — the parents-only
    /// registry loader would return no owns for any mod-plan row.
    #[test]
    fn facts_come_from_ticket_files() {
        let facts = ticket_facts(&worktree_root()).expect("ticket facts");
        assert!(facts.pack_last.contains("T-290"), "T-290 lost pack_last");
        assert_eq!(facts.deps_of("T-212").join(","), "T-685,T-241,T-257");
        assert_eq!(facts.deps_of("T-238").join(","), "T-273,T-237");
        assert_eq!(facts.deps_of("T-251").join(","), "T-209");
        assert!(
            facts.owns.contains_key("T-181.23"),
            "child ticket owns must be globbed, not read through the parents-only loader"
        );
    }
}
